src/Data/Bin/Properties.agda

```diff
@@ -1,143 +1,867 @@
+
+
+------------------------------------------------------------------------------
+-- Constructor properties
+------------------------------------------------------------------------------
+
+x≢0⇒x+y≢0 :  ∀ {x} (y : Bin) → x ≢ 0# → x + y ≢ 0#
+x≢0⇒x+y≢0 {2suc _} 0# _   =  λ()
+x≢0⇒x+y≢0 {0#}     _  0≢0 =  contradiction refl 0≢0
+
+2suc-as∘ :  2suc ≗ 2* ∘ suc
+2suc-as∘ 0#        =  refl
+2suc-as∘ (2suc x)  =  cong 2suc (2suc-as∘ x)
+2suc-as∘ (suc2* x) =  refl
+
+
+-- suc2*∘suc≗suc∘2suc :  suc2* ∘ suc ≗ suc ∘ 2suc   is by  const refl
+
+suc2*-as∘ :  suc2* ≗ suc ∘ 2*
+suc2*-as∘ 0#        = refl
+suc2*-as∘ (2suc x)  = refl
+suc2*-as∘ (suc2* x) = begin
+  suc2* (suc2* x)     ≡⟨ cong suc2* (suc2*-as∘ x) ⟩
+  suc2* (suc 2x)      ≡⟨⟩
+  suc (2suc 2x)       ≡⟨ cong suc (2suc-as∘ 2x) ⟩
+  suc (2* (suc 2x))   ≡⟨ cong (suc ∘ 2*) (sym (suc2*-as∘ x)) ⟩
+  suc (2* (suc2* x))  ∎
+  where
+  2x = 2* x
+
+pred∘suc :  pred ∘ suc ≗ id
+pred∘suc 0#        =  refl
+pred∘suc (2suc x)  =  sym (2suc-as∘ x)
+pred∘suc (suc2* x) =  refl
+
+suc∘pred :  (x : Bin) → x ≢ 0# → suc (pred x) ≡ x
+suc∘pred 0#        0≢0 =  contradiction refl 0≢0
+suc∘pred (2suc _)  _   =  refl
+suc∘pred (suc2* x) _   =  sym (suc2*-as∘ x)
+
+
+------------------------------------------------------------------------------
+-- Addition properties. Chapter I.
+------------------------------------------------------------------------------
+
++-identityˡ : LeftIdentity 0# _+_
++-identityˡ _ = refl
+
++-identityʳ : RightIdentity 0# _+_
++-identityʳ 0#        = refl
++-identityʳ (2suc _)  = refl
++-identityʳ (suc2* _) = refl
+
++-identity : Identity 0# _+_
++-identity = (+-identityˡ , +-identityʳ)
+
+2*≢1 :  {x : Bin} → 2* x ≢ 1B
+2*≢1 {0#} ()
+
+suc≗1+ :  suc ≗ (1B +_)
+suc≗1+ 0#        =  refl
+suc≗1+ (2suc _)  =  refl
+suc≗1+ (suc2* _) =  refl
+
+suc-+ :  ∀ x y → (suc x) + y ≡ suc (x + y)
+suc-+ 0# y =  sym (suc≗1+ y)
+suc-+ x 0# =  begin
+  suc x + 0#     ≡⟨ +-identityʳ (suc x) ⟩
+  suc x          ≡⟨ cong suc (sym (+-identityʳ x)) ⟩
+  suc (x + 0#)   ∎
+
+suc-+ (2suc x) (2suc y)   =  cong (suc ∘ 2suc) (suc-+ x y)
+suc-+ (2suc x) (suc2* y)  =  cong (suc ∘ suc2*) (suc-+ x y)
+suc-+ (suc2* x) (2suc y)  =  refl
+suc-+ (suc2* x) (suc2* y) =  refl
+
+
+1+≗suc :  (1B +_) ≗ suc
+1+≗suc = suc-+ 0#
+
+
+------------------------------------------------------------------------------
+-- Multiplication properties. Chapter I.
+------------------------------------------------------------------------------
+
+*-zeroˡ : LeftZero 0# _*_
+*-zeroˡ _ = refl
+
+*-zeroʳ : RightZero 0# _*_
+*-zeroʳ 0#        = refl
+*-zeroʳ (2suc _)  = refl
+*-zeroʳ (suc2* _) = refl
+
+*-zero : Zero 0# _*_
+*-zero = (*-zeroˡ , *-zeroʳ )
+
+x*y≡0⇒x≡0∨y≡0 :  ∀ x {y} → x * y ≡ 0# → x ≡ 0# ⊎ y ≡ 0#
+x*y≡0⇒x≡0∨y≡0 0# {_}  _ =  inj₁ refl
+x*y≡0⇒x≡0∨y≡0 _  {0#} _ =  inj₂ refl
+
+nz*nz :  ∀ {x y} → x ≢ 0# → y ≢ 0# → x * y ≢ 0#
+nz*nz {x} {_} x≢0 y≢0 xy≡0  with x*y≡0⇒x≡0∨y≡0 x xy≡0
+... | inj₁ x≡0 =  x≢0 x≡0
+... | inj₂ y≡0 =  y≢0 y≡0
+
```

Comment:
This lemma should be called `x≢0∧y≢0⇒x*y≢0`.

Reply:
Renamed